text/000-rustdoc-bundle-local-resources.md

```diff
@@ -0,0 +1,118 @@
+
+Right now, there are attributes that can set a logo and a favicon for documentation, but they must to point to an absolute URL, which prevents bundling the logo and favicon in the source repository. Also, while `<script>` tags are allowed in rustdoc, they have a similar problem: if they load script from some URL, that URL needs to be absolute or it won't work consistently across `cargo doc` and `docs.rs`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC proposes to allow rustdoc to include local images in the generated documentation by copying them into the output directory.
+
+This would be done by allowing users to specify the path of a local resource file in doc comments. The resource file would be stored in the `doc.files` folder. The `doc.files` folder will be at the "top level" of the rustdoc output level (at the same level as the `static.files` or the `src` folders).
+
+The only local resources considered will be the ones in the markdown image syntax: `![resource title](path)`, where `<path>` is the path of the resource file relative to the source file.
+
+The path could be any relative or absolute file path. For example, to include an image generated by [`build.rs`](https://doc.rust-lang.org/cargo/reference/build-scripts.html), concatenate a path with the `OUT_DIR` environment variable:
+
+```rust
+/// Using a local image
+#[doc=concat!("![with absolute path](", env!("OUT_DIR"), "/local/image.png)")]
+///
+/// Using a local image ![with relative path](../local/image.png)
+```
+
+Since the local images are all put in the same folder, if the same is imported from different crates, the content won't be duplicated since they have the same name and the same hash.
+
+If the path isn't referring to a file, a warning will be emitted and rustdoc will left the path unchanged in the generated documentation.
+
+For published crates, `docs.rs` builds the contents of the `.crate` package in a sandbox with no internet access. Make sure any resources your docs need are [included](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) in the package.
+
+The local resources files are not affected by the `--resource-suffix`.
+
+The impact on `docs.rs` would also be very minimal as the size of a published crate resources is limited to a few megabytes. The only thing needed would be to handle the new `doc.files` folder.
```

Comment:
for images this might be true, but I'm a bit worried about JS files. while docs.rs currently allows running arbitrary JS from 3rd-party servers, the situation might become even worse if docs.rs itself would serve those potentially malicious files.

or am I missing something?

Reply:
`![]` syntax generates `<img>` HTML tag and nothing else. So the idea behind this RFC is to bundle the resource (if local) into the generated rustdoc output.

If you put a `.js` file in it, it'll be bundled inside a `<img>` (so not displayed) and I suppose you'll then be able to call it with `<script>` though. But the same can be done currently with any file extension and then load it as JS into a `<script>`. So in this regard, there are no changes I suppose?